larpandoracontent/LArMonitoring/EventClusterValidationAlgorithm.cc

```diff
@@ -0,0 +1,337 @@
+            mainMCParticleHits.at(pMC)++;
+            totalHits++;
+        }
+        if (totalHits == 0) // Something went wrong and won't be able to get metrics from this cluster
+            continue;
+
+        // Determine which MC particle contributes the most weight across the cluster
+        const MCParticle *pMainMC {nullptr};
+        int maxHits {0};
+        for (const auto &[pMC, nHits] : mainMCParticleHits)
+        {
+            if (nHits > maxHits)
+            {
+                pMainMC = pMC;
+                maxHits = nHits;
+            }
+        }
+        metrics.m_purities.push_back(maxHits / totalHits);
+        metrics.m_nRecoHits.push_back(totalHits);
+
+        // Calculate cluster completeness
+        int nTotalMainMCHits {0};
+        for (const auto &[pCaloHit, parents] : hitParents)
+        {
+            if (pMainMC == parents.m_pMainMC)
+                nTotalMainMCHits++;
+        }
+        metrics.m_completenesses.push_back(maxHits / nTotalMainMCHits);
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float EventClusterValidationAlgorithm::CalcRandIndex(std::map<const CaloHit *const, CaloHitParents> &hitParents) const
+{
+    // Fill contingency table
+    ContingencyTable cTable;
+    for (const auto &[pCaloHit, parents] : hitParents)
+    {
+        const MCParticle *const pMC = parents.m_pMainMC;
+        const Cluster *const pCluster = parents.m_pCluster;
+
+        if (cTable.find(pCluster) == cTable.end() || cTable.at(pCluster).find(pMC) == cTable.at(pCluster).end())
+            cTable[pCluster][pMC] = 0;
+        cTable.at(pCluster).at(pMC)++;
+    }
+
+    return LArMonitoringHelper::CalcRandIndex(cTable);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void EventClusterValidationAlgorithm::SetBranches(ClusterMetrics &metrics, float randIndex, std::string branchPrefix) const
+{
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName, branchPrefix + "n_hits", metrics.m_nHits ));
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName, branchPrefix + "n_clusters", metrics.m_nClusters));
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName, branchPrefix + "n_mainMCs", metrics.m_nMainMCs));
+    double meanPurity = -1.0, meanCompleteness = -1.0, meanNRecoHits = -1.0;
+    if (metrics.m_nClusters > 0)
+    {
+        meanPurity = std::accumulate(metrics.m_purities.begin(), metrics.m_purities.end(), 0.0) / metrics.m_purities.size();
+        meanCompleteness = std::accumulate(metrics.m_completenesses.begin(), metrics.m_completenesses.end(), 0.0) / metrics.m_completenesses.size();
+        meanNRecoHits = std::accumulate(metrics.m_nRecoHits.begin(), metrics.m_nRecoHits.end(), 0.0) / metrics.m_nRecoHits.size();
```

Comment:
I couldn't quite convince myself that the denomenator here can never be zero.  I think it's probably ok, but can the code get here if hitParents ends up being empty?

Reply:
I didnt have any problems with this on a 50k sample but I think your suspicion is correct as if truth matching was completely missing hitParents would be empty. I just checked everything to be safe